saleor/page/models.py

```diff
@@ -0,0 +1,47 @@
+import datetime
+
+from django.core.validators import validate_slug
+from django.db import models
+from django.db.models import Q
+from django.urls import reverse
+
+
+class PageManager(models.QuerySet):
+    def public(self):
+        today = datetime.date.today()
+        return self.filter(
+            Q(visible=True),
+            Q(available_on__lte=today) | Q(available_on__isnull=True))
+
+    def get_available(self, allow_draft=False):
+        if not allow_draft:
+            return self.public()
+        return self
+
+
+class Page(models.Model):
+    url = models.CharField(
+        max_length=100, db_index=True, validators=[validate_slug])
+    title = models.CharField(max_length=200)
+    content = models.TextField()
+    created = models.DateTimeField(auto_now_add=True)
+    visible = models.BooleanField(default=False)
+    available_on = models.DateField(blank=True, null=True)
+
+    objects = PageManager.as_manager()
+
+    class Meta:
+        ordering = ('url',)
+
+    def __str__(self):
+        return self.title
+
+    def get_absolute_url(self):
+        return reverse('page:details', kwargs={'url': self.url})
+
+    def save(self, *args, **kwargs):
+        """
+        Make sure url is not being written to database with uppercase.
+        """
+        self.url = self.url.lower()
```

Comment:
In my opinion, this is part of validation that should be handled by a form.

Reply:
I was considering this, but decided to put it there and I think it's better. I can imagine developer playing around in shell and creating page with invalid URL, which is being written to database and causing long term problems.

Reply:
I agree with @elwoodxblues, `save()` is not a good place for logic. Having it modify the instance is an unexpected side effect.